src/conn_poll.c

```diff
@@ -141,299 +313,487 @@ int conn_poll_prepare(conn_registry_t *registry, pfds_record_t *pfds, timestamp_
+	connection->send_offset = 0;
+	connection->header_bytes = 0;
+	connection->frame_len = 0;
+	connection->payload_len = 0;
+
+	addr_unmap_inet6_v4mapped((struct sockaddr *)&connection->remote.addr, &connection->remote.len);
+
+	if (!connecting)
+		JLOG_DEBUG("ICE-TCP active connection established");
+	else
+		JLOG_DEBUG("ICE-TCP active connection in progress");
+
+	if (agent_conn_tcp_state(agent, &connection->remote,
+	                        connecting ? TCP_STATE_CONNECTING : TCP_STATE_CONNECTED) != 0) {
+		conn_poll_free_tcp_connection(connection);
+		closesocket(sock);
+		return NULL;
 	}
-	memset(&conn_impl->tcp_ice_write_context, 0, sizeof(tcp_ice_write_context_t));
-	memset(&conn_impl->tcp_ice_read_context, 0, sizeof(tcp_ice_read_context_t));
-	conn_poll_change_tcp_state(agent, TCP_STATE_FAILED);
+
+	++conn_impl->tcp_conns_count;
+	return connection;
 }
 
-void conn_poll_change_tcp_state(juice_agent_t *agent, tcp_state_t state) {
-	conn_impl_t *conn_impl = agent->conn_impl;
-	if(conn_impl->tcp_state != state) {
-		switch(state) {
-			case TCP_STATE_DISCONNECTED:
-				JLOG_DEBUG("TCP state changed to disconnected");
-				break;
-			case TCP_STATE_CONNECTING:
-				JLOG_DEBUG("TCP state changed to connecting");
-				break;
-			case TCP_STATE_CONNECTED:
-				JLOG_DEBUG("TCP state changed to connected");
-				break;
-			case TCP_STATE_FAILED:
-				JLOG_DEBUG("TCP state changed to failed");
-				break;
-			default:
-				break;
+static int conn_poll_tcp_handle_read(juice_agent_t *agent, tcp_connection_t *connection) {
+	uint8_t buffer[BUFFER_SIZE];
+	for (;;) {
+        int ret = (int)recv(connection->sock, (char *)buffer, sizeof(buffer), 0);
+		if (ret < 0) {
+			if (sockerrno == SEAGAIN || sockerrno == SEWOULDBLOCK)
+				return 0;
+			JLOG_WARN("TCP recv failed, errno=%d", sockerrno);
+			return -1;
 		}
-		conn_impl->tcp_state = state;
-		if (agent_conn_tcp_state(agent, &conn_impl->tcp_dst, state) != 0) {
-			if (conn_impl->tcp_sock != INVALID_SOCKET) {
-				closesocket(conn_impl->tcp_sock);
-				conn_impl->tcp_sock = INVALID_SOCKET;
+		if (ret == 0)
+			return -1; // closed
+
+		size_t offset = 0;
+		while (offset < (size_t)ret) {
+			if (connection->header_bytes < 2) {
+				connection->frame_len = (uint16_t)((connection->frame_len << 8) | buffer[offset]);
+				++connection->header_bytes;
+				++offset;
+				if (connection->header_bytes == 2) {
+					if (connection->frame_len == 0) {
+						JLOG_WARN("Invalid zero-length TCP frame");
+						return -1;
+					}
+					if (connection->frame_len >= sizeof(connection->recv_buffer)) {
+						JLOG_WARN("TCP frame too large: %u bytes", connection->frame_len);
+						return -1;
+					}
+					connection->payload_len = 0;
+				}
+				continue;
+			}
+
+			size_t remaining = connection->frame_len - connection->payload_len;
+			size_t copy = remaining < ((size_t)ret - offset) ? remaining : ((size_t)ret - offset);
+			memcpy(connection->recv_buffer + connection->payload_len, buffer + offset, copy);
```

Comment:
Everything is inefficiently copied here while you could instead receive in-place in the buffer.

This was already dome by `tcp_ice_read()`, please use the existing function.